go/vt/vtgate/engine/mirror.go

```diff
@@ -0,0 +1,219 @@
+		defer mirrorCtxCancel()
+
+		go func(target Primitive, vcursor VCursor) {
+			defer close(mirrorCn)
+			_ = target.TryStreamExecute(mirrorCtx, vcursor, bindVars, wantfields, func(_ *sqltypes.Result) error {
+				return nil
+			})
+		}(m.Target, vcursor.CloneForMirroring(mirrorCtx))
+	}
+
+	err := m.Primitive.TryStreamExecute(ctx, vcursor, bindVars, wantfields, callback)
+
+	if mirrorCn != nil {
+		select {
+		case <-mirrorCn:
+		case <-time.After(maxMirrorTargetLag):
+			mirrorCtxCancel()
+		}
+	}
+
+	return err
+}
+
+// Inputs is a slice containing the inputs to this Primitive.
+// The returned map has additional information about the inputs, that is used in the description.
+func (m *Mirror) Inputs() ([]Primitive, []map[string]any) {
+	return []Primitive{m.Primitive, m.Target}, nil
+}
+
+// description is the description, sans the inputs, of this Primitive.
+// to get the plan description with all children, use PrimitiveToPlanDescription()
+func (m *Mirror) description() PrimitiveDescription {
+	return PrimitiveDescription{
+		OperatorType: "Mirror",
+	}
+}
+
+func (m *PercentMirrorTarget) RouteType() string {
+	return "PercentMirrorTarget"
+}
+
+func (m *PercentMirrorTarget) GetKeyspaceName() string {
+	return m.Primitive.GetKeyspaceName()
+}
+
+func (m *PercentMirrorTarget) GetTableName() string {
+	return m.Primitive.GetTableName()
+}
+
+func (m *PercentMirrorTarget) GetFields(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return m.Primitive.GetFields(ctx, vcursor, bindVars)
+}
+
+func (m *PercentMirrorTarget) NeedsTransaction() bool {
+	return m.Primitive.NeedsTransaction()
+}
+
+func (m *PercentMirrorTarget) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	return m.Primitive.TryExecute(ctx, vcursor, bindVars, wantfields)
+}
+
+func (m *PercentMirrorTarget) TryStreamExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	return m.Primitive.TryStreamExecute(ctx, vcursor, bindVars, wantfields, callback)
+}
+
+// Inputs is a slice containing the inputs to this Primitive.
+// The returned map has additional information about the inputs, that is used in the description.
+func (m *PercentMirrorTarget) Inputs() ([]Primitive, []map[string]any) {
+	return []Primitive{m.Primitive}, nil
+}
+
+// description is the description, sans the inputs, of this Primitive.
+// to get the plan description with all children, use PrimitiveToPlanDescription()
+func (m *PercentMirrorTarget) description() PrimitiveDescription {
+	return PrimitiveDescription{
+		OperatorType: "MirrorTarget",
+		Variant:      "Percent",
+		Other: map[string]any{
+			"Percent": m.Percent,
+		},
+	}
+}
+
+func (m *PercentMirrorTarget) Accept() bool {
+	return m.Percent >= (rand.Float32() * 100.0)
```

Comment:
Feels a little heavyweight that `PercentMirrorTarget` implements `Primitive` like this. Couldn't we just store the m.Percent and the Primitive as two different fields in the `Mirror` struct? A lot of pass-through methods here that we could avoid

Reply:
the one downside to this is the route descriptions (in `vexplain` for example) won't clearly label which primitive is the mirror source, and which is the target, but i'm fine w/ it if you are.

Reply:
You can add additional information to the children. Check this out:

```golang
	Primitive interface {
	<...>
		// The returned map has additional information about the inputs, that is used in the description.
		Inputs() ([]Primitive, []map[string]any)
	<...>
	}
```

`FkCascade` uses it to add more info